// playwright-core/lib/api.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var accessibility_1 = require("./accessibility");
exports.Accessibility = accessibility_1.Accessibility;
var browserContext_1 = require("./browserContext");
exports.BrowserContext = browserContext_1.BrowserContext;
var console_1 = require("./console");
exports.ConsoleMessage = console_1.ConsoleMessage;
var dialog_1 = require("./dialog");
exports.Dialog = dialog_1.Dialog;
var dom_1 = require("./dom");
exports.ElementHandle = dom_1.ElementHandle;
var errors_1 = require("./errors");
exports.TimeoutError = errors_1.TimeoutError;
var frames_1 = require("./frames");
exports.Frame = frames_1.Frame;
var input_1 = require("./input");
exports.Keyboard = input_1.Keyboard;
exports.Mouse = input_1.Mouse;
var javascript_1 = require("./javascript");
exports.JSHandle = javascript_1.JSHandle;
var network_1 = require("./network");
exports.Request = network_1.Request;
exports.Response = network_1.Response;
var page_1 = require("./page");
exports.Page = page_1.Page;
exports.Worker = page_1.Worker;
var selectors_1 = require("./selectors");
exports.Selectors = selectors_1.Selectors;
var crBrowser_1 = require("./chromium/crBrowser");
exports.ChromiumBrowser = crBrowser_1.CRBrowser;
var crCoverage_1 = require("./chromium/crCoverage");
exports.ChromiumCoverage = crCoverage_1.CRCoverage;
var crConnection_1 = require("./chromium/crConnection");
exports.ChromiumSession = crConnection_1.CRSession;
var crTarget_1 = require("./chromium/crTarget");
exports.ChromiumTarget = crTarget_1.CRTarget;
var ffBrowser_1 = require("./firefox/ffBrowser");
exports.FirefoxBrowser = ffBrowser_1.FFBrowser;
var wkBrowser_1 = require("./webkit/wkBrowser");
exports.WebKitBrowser = wkBrowser_1.WKBrowser;
var browserServer_1 = require("./server/browserServer");
exports.BrowserServer = browserServer_1.BrowserServer;
